### PR TITLE
[Feature] Add Orc Chunk Writer

### PR DESCRIPTION
**Pull request by YinZheng-Sun**
Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exec/orc_writer.cpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    ASSIGN_OR_RETURN(auto writable_file, _fs->new_writable_file(options, new_file_location))
+
+    _writer = std::make_shared<AsyncOrcChunkWriter>(std::move(writable_file), _options, _schema, _output_expr_ctxs,
+                                                    ExecEnv::GetInstance()->pipeline_sink_io_pool(), _parent_profile);
+
+    // auto st = _writer->init();
+    return Status::OK();
+}
+
+Status RollingAsyncOrcWriter::append_chunk(Chunk* chunk, RuntimeState* state) {
```
Comment (GavinMar) [inline be/src/exec/orc_writer.cpp:58]: Where the chunk is appened?

Comment (YinZheng-Sun) [reply be/src/exec/orc_writer.cpp:57]: It is a mistake, sorry about that.

File: be/src/exec/orc_writer.cpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    if (_writer == nullptr) {
+        auto status = _new_chunk_writer();
+        if (!status.ok()) {
```
[the remark this answers is not in the document]
Comment (YinZheng-Sun) [reply be/src/exec/orc_writer.cpp:60]: OK

File: be/src/formats/orc/orc_chunk_writer.cpp [hunk 3]
```diff
@@ -0,0 +1,746 @@
...
+    }
+
+    root.numElements = num_rows;
+    RETURN_IF_ERROR(_flush_batch());
+    return Status::OK();
+}
+
+Status OrcChunkWriter::_flush_batch() {
+    if (!_writer) {
+        return Status::InternalError("ORC Writer is not inited");
```
[the remark this answers is not in the document]
Comment (YinZheng-Sun) [reply be/src/formats/orc/orc_chunk_writer.cpp]: Changed

File: be/src/formats/orc/orc_chunk_writer.cpp [hunk 4]
```diff
@@ -0,0 +1,739 @@
...
+    number_orc_column.resize(column_size);
+    number_orc_column.notNull.resize(column_size);
+
+    if (column->is_nullable()) {
+        auto c = ColumnHelper::as_raw_column<NullableColumn>(column);
+        auto* nulls = c->null_column()->get_data().data();
+        auto* values = ColumnHelper::cast_to_raw<Type>(c->data_column())->get_data().data();
+
+        for (size_t i = 0; i < column_size; ++i) {
+            if (nulls[i]) {
```
Comment (letian-jiang) [inline be/src/formats/orc/orc_chunk_writer.cpp:324]: Is it possible to remove this in-loop branch (to enable auto-vectorization)? 

I guess it's ok to populate `number_orc_column[i]` even if `nulls[i] != 0`.

Comment (YinZheng-Sun) [reply be/src/formats/orc/orc_chunk_writer.cpp]: Do you mean do redundant copy for auto-vectorization? I think it is an good idea and have changed.

**Review by letian-jiang — APPROVED** [review]
LGTM

File: be/src/formats/orc/orc_chunk_writer.cpp [hunk 5]
```diff
@@ -0,0 +1,736 @@
...
+        auto* values = reinterpret_cast<int128_t*>(down_cast<DecimalColumn*>(column.get())->get_data().data());
+        for (size_t i = 0; i < column_size; ++i) {
+            decimal_orc_column.values[i] = orc::Int128(values[i] >> 64, (values[i] << 64) >> 64);
+        }
+        memset(decimal_orc_column.notNull.data(), 1, column_size * sizeof(char));
+    }
+    decimal_orc_column.numElements = column_size;
+}
+
+void OrcChunkWriter::_write_datetimes(orc::ColumnVectorBatch& orc_column, ColumnPtr& column) {
```
Comment (GavinMar) [inline be/src/formats/orc/orc_chunk_writer.cpp:464]: You can use the name `_write_date` to ensure that function and type naming are consistent.

File: be/src/formats/orc/orc_chunk_writer.cpp [hunk 6]
```diff
@@ -0,0 +1,736 @@
...
+        auto* values = ColumnHelper::cast_to_raw<TYPE_DATE>(column)->get_data().data();
+        for (size_t i = 0; i < column_size; ++i) {
+            date_orc_column.data[i] = OrcDateHelper::native_date_to_orc_date(values[i]);
+        }
+        memset(date_orc_column.notNull.data(), 1, column_size * sizeof(char));
+    }
+    date_orc_column.numElements = column_size;
+}
+
+void OrcChunkWriter::_write_timestamps(orc::ColumnVectorBatch& orc_column, ColumnPtr& column) {
```
Comment (GavinMar) [inline be/src/formats/orc/orc_chunk_writer.cpp:495]: _write_timestamps ->  _write_datatime

**Review by GavinMar — APPROVED** [review]

**Comment by YinZheng-Sun** [comment]
updated

**Review by dirtysalt — APPROVED** [review]

File: be/src/formats/orc/orc_chunk_writer.cpp [hunk 7]
```diff
@@ -0,0 +1,736 @@
...
+
+Status OrcChunkWriter::_write_struct_column(orc::ColumnVectorBatch& orc_column, ColumnPtr& column,
+                                            const TypeDescriptor& type) {
+    auto& struct_orc_column = dynamic_cast<orc::StructVectorBatch&>(orc_column);
+    auto column_size = column->size();
+
+    struct_orc_column.resize(column_size);
+    struct_orc_column.notNull.resize(column_size);
+
+    if (column->is_nullable()) {
```
Comment (stephen-shelby) [inline be/src/formats/orc/orc_chunk_writer.cpp:575]: Maybe you need to extract the same logic to make the code more readable in the later PR.

File: be/src/formats/orc/orc_chunk_writer.cpp [hunk 8]
```diff
@@ -0,0 +1,736 @@
...
+    }
+    case TCompressionType::ZSTD: {
+        _writer_options->setCompression(orc::CompressionKind::CompressionKind_ZSTD);
+        break;
+    }
+    case TCompressionType::LZ4: {
+        _writer_options->setCompression(orc::CompressionKind::CompressionKind_LZ4);
+        break;
+    }
+    case TCompressionType::LZO: {
```
Comment (stephen-shelby) [inline be/src/formats/orc/orc_chunk_writer.cpp:102]: does orc support lzo?

**Review by stephen-shelby — APPROVED** [review]

File: be/src/formats/orc/orc_chunk_writer.cpp (hunk 8 above)
Comment (YinZheng-Sun) [reply be/src/formats/orc/orc_chunk_writer.cpp:102]: It seems that orc only supports decompress from lzo and doesn't support compress to lzo. I will fix it in the next pr.

File: be/src/formats/orc/orc_chunk_writer.cpp (hunk 7 above)
Comment (YinZheng-Sun) [reply be/src/formats/orc/orc_chunk_writer.cpp:575]: Okay

**Pull request merged**